File: ServerPhp/src/dc/DC0024.php
<?php
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0024 Controller: Accounting period
 */

require_once(PATH_N21LIB."/Http/Controller.php");

class DC0024 extends Controller {

  public function __construct($request, $response, $session) {
    parent::__construct($request, $response, $session);
    $this->_initFields();
  }

private function preQuery(&$params, &$where) {
    if (!empty($_REQUEST["QRY_ACCYEAR_CODE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.ACCYEAR_CODE like :ACCYEAR_CODE";
      $params["ACCYEAR_CODE"] = $_REQUEST["QRY_ACCYEAR_CODE"];
    }
    if (!empty($_REQUEST["QRY_CLIENT_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CLIENT_ID like :CLIENT_ID";
      $params["CLIENT_ID"] = $_REQUEST["QRY_CLIENT_ID"];
    }
    if (!empty($_REQUEST["QRY_CLOSED"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CLOSED like :CLOSED";
      $params["CLOSED"] = $_REQUEST["QRY_CLOSED"];
    }
    if (!empty($_REQUEST["QRY_CODE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CODE like :CODE";
      $params["CODE"] = $_REQUEST["QRY_CODE"];
      $params["CODE"] = strtoupper($params["CODE"]);
    }
    if (!empty($_REQUEST["QRY_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.ID like :ID";
      $params["ID"] = $_REQUEST["QRY_ID"];
    }
}

public function doQuery() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                t.ACCYEAR_CODE
                ,pbo_client.get_code_by_id(t.client_id) CLIENT_CODE
                ,t.CLIENT_ID
                ,t.CLOSED
                ,t.CODE
                ,t.CREATEDBY
                ,t.CREATEDON
                ,t.ENDDATE
                ,t.ID
                ,t.IS_FIRST_PERIOD
                ,t.MODIFIEDBY
                ,t.MODIFIEDON
                ,t.NAME
                ,t.NOTES
                ,t.OPENED
                ,t.PERIODTYPE
                ,t.PREV_PERIOD_NAME
                ,t.PROCESSING
                ,t.STARTDATE
            from AC_ACC_PERIOD t $where $orderByClause ";
    $this->logger->debug($sql);
    $rs = $this->getDbConn()->SelectLimit($sql, $limit, $start, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $dataOut = $this->serializeCursor($rs,array_keys($this->fields), $this->getDataFormatFetch());
    if ($this->getDataFormatFetch() == "xml" ) {header("Content-type: application/xml");}
    print $this->formatQueryResponseData($dataOut, $rsCount->fields["TOTALCOUNT"]);
} /* end function doQuery  */


public function doExport() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $groupBy = (!empty($_REQUEST["groupBy"]))?$_REQUEST["groupBy"]:"";
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                t.ID
                ,t.CLIENT_ID
                ,pbo_client.get_code_by_id(t.client_id) CLIENT_CODE
                ,t.ACCYEAR_CODE
                ,t.CODE
                ,t.NAME
                ,t.STARTDATE
                ,t.ENDDATE
                ,t.PREV_PERIOD_NAME
                ,t.OPENED
                ,t.CLOSED
                ,t.NOTES
                ,t.IS_FIRST_PERIOD
                ,t.CREATEDON
                ,t.CREATEDBY
                ,t.MODIFIEDON
                ,t.MODIFIEDBY
                ,t.PERIODTYPE
                ,t.PROCESSING
            from AC_ACC_PERIOD t $where $orderByClause ";
    $rs = $this->getDbConn()->Execute($sql, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    if (!empty($_REQUEST["_p_disp_cols"])) {
      $columns = explode("|",$_REQUEST["_p_disp_cols"]);
    } else {
      $columns = array_keys($this->fields);
    }
    if ($this->getDataFormatExport() == "csv" ) {
      $dataOut = $this->serializeCursor($rs,$columns,"csv");
    } else {
      $dataOut = $this->serializeCursor($rs,$columns,"xml");
      $dataOut = "<records>".$dataOut."</records>";
      $dataOut = "<queryParams>".$this->serializeMap($params,"xml")."</queryParams>".$dataOut;
      $dataOut = "<columnDef>".$this->columnDefForExport($columns,$this->fields,true).$this->columnDefForExport(array_diff(array_keys($params), $columns),$this->fields,false)."</columnDef>".$dataOut;
      $dataOut = "<staticText>".$this->exportLocalizedStaticText()."</staticText>".$dataOut;
      $dataOut = "<groupBy>".$groupBy."</groupBy>".$dataOut;
      $dataOut = "<reportData  title=\"".$this->getDcTitle()."\" by=\"".$_SESSION["user"]["userName"]."\" on=\"".date(DATE_FORMAT)."\">".$dataOut."</reportData>";
    }
    $this->beginExport();
    print $dataOut;
    $this->endExport();
} /* end function doExport  */

public function fetchRecord() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0024.fetchRecord().");}
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function fetchRecord */

public function doInsert() {
    $this->populateRecordFromRequest(); 
    $sql = "insert into AC_ACC_PERIOD(
                 ACCYEAR_CODE
                ,CLIENT_ID
                ,CODE
                ,CREATEDBY
                ,ENDDATE
                ,ID
                ,IS_FIRST_PERIOD
                ,MODIFIEDBY
                ,NOTES
                ,PERIODTYPE
                ,PREV_PERIOD_NAME
                ,PROCESSING
                ,STARTDATE
            ) values ( 
                 :ACCYEAR_CODE
                ,:CLIENT_ID
                ,:CODE
                ,:CREATEDBY
                ,:ENDDATE
                ,:ID
                ,:IS_FIRST_PERIOD
                ,:MODIFIEDBY
                ,:NOTES
                ,:PERIODTYPE
                ,:PREV_PERIOD_NAME
                ,:PROCESSING
                ,:STARTDATE
    )";
    $stmt = $this->getDbConn()->prepare($sql);
    $_seq = $this->getDbConn()->execute("select seq_accperiod_id.nextval seq_val from dual")->fetchRow();
    $this->record["ID"] = $_seq["SEQ_VAL"];
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doInsert */

public function doUpdate() {
    $this->populateRecordFromRequest();
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0024.doUpdate().");}
    $sql = "update AC_ACC_PERIOD set 
                 ID=:ID
                ,CODE=:CODE
                ,STARTDATE=:STARTDATE
                ,ENDDATE=:ENDDATE
                ,PREV_PERIOD_NAME=:PREV_PERIOD_NAME
                ,NOTES=:NOTES
                ,IS_FIRST_PERIOD=:IS_FIRST_PERIOD
    where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"],"ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doUpdate */

public function doDelete() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0024.doDelete().");}
    $sql = "delete from AC_ACC_PERIOD where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doDelete */

public function initNewRecord() {
    $this->populateRecordFromRequest();
    $this->record["_p_record_status"] = "insert";
    $this->setFieldInitialValues($this->record, "DC0024");
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function initNewRecord */

private function findByPk(&$pkCols, &$record) {
    $sql = "select 
                t.ACCYEAR_CODE
                ,pbo_client.get_code_by_id(t.client_id) CLIENT_CODE
                ,t.CLIENT_ID
                ,t.CLOSED
                ,t.CODE
                ,t.CREATEDBY
                ,t.CREATEDON
                ,t.ENDDATE
                ,t.ID
                ,t.IS_FIRST_PERIOD
                ,t.MODIFIEDBY
                ,t.MODIFIEDON
                ,t.NAME
                ,t.NOTES
                ,t.OPENED
                ,t.PERIODTYPE
                ,t.PREV_PERIOD_NAME
                ,t.PROCESSING
                ,t.STARTDATE
            from AC_ACC_PERIOD t
         where 
           t.ID= :ID
            ";
    $rs = $this->getDbConn()->Execute($sql, $pkCols);
    $row = $rs->FetchRow();
    CollectionUtils::copyMapValues($record, $row);
} /* end function findByPk  */


public function callProcedure($pName) {
  $this->logger->debug("Start: ".$this->dcName.".callProcedure(".$pName.")");
    $this->populateRecordFromRequest();
    if ($pName == "OpenAccPeriod") { $this->callProc_OpenAccPeriod(); }
    print "{success:true, data:".json_encode($this->record)."}";
    $this->logger->debug("End: ".$this->dcName.".callProcedure");
} /* end function callProcedure */


private function callProc_OpenAccPeriod() {
  $this->logger->debug("Start: ".$this->dcName.".callProc_OpenAccPeriod");


  $stmt = $this->getDbConn()->PrepareSP("BEGIN ps_acc.open_accperiod(p_accperiod_id => :p_accperiod_id); END;");
  $this->getDbConn()->InParameter($stmt,$this->record["ID"] ,"p_accperiod_id" );
  $this->logger->debug("In-Parameters: ".CollectionUtils::mapToString($this->record) );
  $this->getDbConn()->Execute($stmt);
  $this->logger->debug("Out-Parameters: ".CollectionUtils::mapToString($this->record) );

  $this->logger->debug("End: ".$this->dcName.".callProc_OpenAccPeriod");
} /* end function callProc_OpenAccPeriod  */


private function _initFields() {
  $this->fields = array(
  "ACCYEAR_CODE" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CLIENT_CODE" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CLIENT_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"CLOSED" => array(parent::FLDPROP_DATA_TYPE => "BOOLEAN")
  ,"CODE" => array(parent::FLDPROP_DATA_TYPE => "STRING",parent::FLDPROP_CASE_RESTRICTION => "Upper")
  ,"CREATEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CREATEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"ENDDATE" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"IS_FIRST_PERIOD" => array(parent::FLDPROP_DATA_TYPE => "BOOLEAN")
  ,"MODIFIEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"MODIFIEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"NOTES" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"OPENED" => array(parent::FLDPROP_DATA_TYPE => "BOOLEAN")
  ,"PERIODTYPE" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"PREV_PERIOD_NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"PROCESSING" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"STARTDATE" => array(parent::FLDPROP_DATA_TYPE => "DATE")
);
}

}
?>

File: ServerJava/src/net/nan21/ebs/dc/DC0024.java
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0024 DC Controller: Accounting period
 */

package net.nan21.ebs.dc;


import java.util.*;
import javax.servlet.http.HttpServletResponse;
import net.nan21.lib.*;

public class DC0024 extends AbstractDataControl implements IDataControl {

  public void init (HttpRequest request, HttpServletResponse response, HttpSession session, DbManager dbm) throws Exception {
    this._initFields();
    super.init(request, response, session, dbm);
  }

private void preQuery() {
    if (this.request.getParam("QRY_ID") != null && !this.request.getParam("QRY_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.ID like :ID");
      this.queryParams.put("ID",(String)this.request.getParam("QRY_ID"));
    }
    if (this.request.getParam("QRY_CLIENT_ID") != null && !this.request.getParam("QRY_CLIENT_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CLIENT_ID like :CLIENT_ID");
      this.queryParams.put("CLIENT_ID",(String)this.request.getParam("QRY_CLIENT_ID"));
    }
    if (this.request.getParam("QRY_ACCYEAR_CODE") != null && !this.request.getParam("QRY_ACCYEAR_CODE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.ACCYEAR_CODE like :ACCYEAR_CODE");
      this.queryParams.put("ACCYEAR_CODE",(String)this.request.getParam("QRY_ACCYEAR_CODE"));
    }
    if (this.request.getParam("QRY_CODE") != null && !this.request.getParam("QRY_CODE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CODE like :CODE");
      this.queryParams.put("CODE",(String)this.request.getParam("QRY_CODE").toUpperCase());
    }
    if (this.request.getParam("QRY_CLOSED") != null && !this.request.getParam("QRY_CLOSED").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CLOSED like :CLOSED");
      this.queryParams.put("CLOSED",(String)this.request.getParam("QRY_CLOSED"));
    }
}

public void doQuery() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " t.ID"+
               " ,t.CLIENT_ID"+
               " ,t.ACCYEAR_CODE"+
               " ,t.CODE"+
               " ,t.STARTDATE"+
               " ,t.ENDDATE"+
               " ,t.CLOSED"+
               " ,t.CREATEDON"+
               " ,t.CREATEDBY"+
               " ,t.MODIFIEDON"+
               " ,t.MODIFIEDBY"+
               " ,t.PREV_PERIOD_NAME"+
               " ,t.NOTES"+
               " ,t.PERIODTYPE"+
               " ,t.PROCESSING"+
               " ,t.OPENED"+
               " ,pbo_client.get_code_by_id(t.client_id) CLIENT_CODE"+
               " ,t.NAME"+
               " ,t.IS_FIRST_PERIOD"+
           " from AC_ACC_PERIOD t "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoQuery(sql);
} 


public void doExport() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " t.ID"+
               " ,t.CLIENT_ID"+
               ",pbo_client.get_code_by_id(t.client_id) CLIENT_CODE"+
               " ,t.ACCYEAR_CODE"+
               " ,t.CODE"+
               " ,t.NAME"+
               " ,t.STARTDATE"+
               " ,t.ENDDATE"+
               " ,t.PREV_PERIOD_NAME"+
               " ,t.OPENED"+
               " ,t.CLOSED"+
               " ,t.NOTES"+
               " ,t.IS_FIRST_PERIOD"+
               " ,t.CREATEDON"+
               " ,t.CREATEDBY"+
               " ,t.MODIFIEDON"+
               " ,t.MODIFIEDBY"+
               " ,t.PERIODTYPE"+
               " ,t.PROCESSING"+
           " from AC_ACC_PERIOD t "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoExport(sql);
}

 	public void fetchRecord()  throws Exception  {
     this.populateRecordPkFromRequest();
     this.findByPk();
       this.writeResultFetchRecord();	 
   }

public void doInsert()  throws Exception {
  this.populateRecordFromRequest(); 
  this.populateRecordWithClientSpecific();
    String sql = "insert into AC_ACC_PERIOD("+
               "  ID"+
               " ,CLIENT_ID"+
               " ,ACCYEAR_CODE"+
               " ,CODE"+
               " ,STARTDATE"+
               " ,ENDDATE"+
               " ,CREATEDBY"+
               " ,MODIFIEDBY"+
               " ,PREV_PERIOD_NAME"+
               " ,NOTES"+
               " ,PERIODTYPE"+
               " ,PROCESSING"+
               " ,IS_FIRST_PERIOD"+
           " ) values ( "+
               "  :ID"+
               " ,:CLIENT_ID"+
               " ,:ACCYEAR_CODE"+
               " ,:CODE"+
               " ,:STARTDATE"+
               " ,:ENDDATE"+
               " ,:CREATEDBY"+
               " ,:MODIFIEDBY"+
               " ,:PREV_PERIOD_NAME"+
               " ,:NOTES"+
               " ,:PERIODTYPE"+
               " ,:PROCESSING"+
               " ,:IS_FIRST_PERIOD"+
    ")";
    this.record.put("ID",   dbm.getSequenceNextValue("seq_accperiod_id")  );
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoInsert();
}

public void doUpdate() throws Exception {
    this.populateRecordFromRequest();
    this.populateRecordWithClientSpecific();
    String sql = "update AC_ACC_PERIOD set "+
               "  ID=:ID"+
               " ,CODE=:CODE"+
               " ,STARTDATE=:STARTDATE"+
               " ,ENDDATE=:ENDDATE"+
               " ,PREV_PERIOD_NAME=:PREV_PERIOD_NAME"+
               " ,NOTES=:NOTES"+
               " ,IS_FIRST_PERIOD=:IS_FIRST_PERIOD"+
   " where "+
     "      ID= :ID"+
   "";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoUpdate();	
}

public void doDelete() throws Exception {
    this.populateRecordPkFromRequest();
   String sql = "delete from AC_ACC_PERIOD where "+
     "      ID= :ID"+
    "";
    dbm.executeStatement(sql, this.recordPk);
   this.writeResultDoDelete();
}

public void initNewRecord() throws Exception {
    this.populateRecordFromRequest();
    this.record.put("_p_record_status", "insert");
    this.writeResultInitNewRecord();
}

private void findByPk()  throws Exception {
    String sql = "select "+ 
               " t.ID"+
               " ,t.CLIENT_ID"+
               " ,t.ACCYEAR_CODE"+
               " ,t.CODE"+
               " ,t.STARTDATE"+
               " ,t.ENDDATE"+
               " ,t.CLOSED"+
               " ,t.CREATEDON"+
               " ,t.CREATEDBY"+
               " ,t.MODIFIEDON"+
               " ,t.MODIFIEDBY"+
               " ,t.PREV_PERIOD_NAME"+
               " ,t.NOTES"+
               " ,t.PERIODTYPE"+
               " ,t.PROCESSING"+
               " ,t.OPENED"+
                ",pbo_client.get_code_by_id(t.client_id) CLIENT_CODE"+
               " ,t.NAME"+
               " ,t.IS_FIRST_PERIOD"+
           " from AC_ACC_PERIOD t"+
        " where "+
     "      t.ID= :ID"+ 
          "";
    this.record = dbm.executeQuery(sql, this.recordPk).get(0); 
} 


public void doCustomAction(String pName)  throws Exception {
    this.populateRecordFromRequest();
    if (pName.equals("OpenAccPeriod")) { this.callProc_OpenAccPeriod(); }
}


private void callProc_OpenAccPeriod() throws Exception {


  Properties inParams = new Properties();
  Properties outParams = new Properties();
   inParams.setProperty("p_accperiod_id", "ID");
  String sql = "BEGIN ps_acc.open_accperiod(?); END;";
  dbm.executeProcedure(sql, inParams, outParams, this.record);

} 


	private void  _initFields() {
	  this.fields = new HashMap<String, FieldDef>();
	  this.fields.put("ID", new FieldDef("NUMBER"));
	  this.fields.put("CLIENT_ID", new FieldDef("NUMBER"));
	  this.fields.put("ACCYEAR_CODE", new FieldDef("STRING"));
	  this.fields.put("CODE", new FieldDef("STRING"));
	  this.fields.get("CODE").setCaseRestriction("Upper");
	  this.fields.put("STARTDATE", new FieldDef("DATE"));
	  this.fields.put("ENDDATE", new FieldDef("DATE"));
	  this.fields.put("CLOSED", new FieldDef("BOOLEAN"));
	  this.fields.put("CREATEDON", new FieldDef("DATE"));
	  this.fields.put("CREATEDBY", new FieldDef("STRING"));
	  this.fields.put("MODIFIEDON", new FieldDef("DATE"));
	  this.fields.put("MODIFIEDBY", new FieldDef("STRING"));
	  this.fields.put("PREV_PERIOD_NAME", new FieldDef("STRING"));
	  this.fields.put("NOTES", new FieldDef("STRING"));
	  this.fields.put("PERIODTYPE", new FieldDef("STRING"));
	  this.fields.put("PROCESSING", new FieldDef("STRING"));
	  this.fields.put("OPENED", new FieldDef("BOOLEAN"));
	  this.fields.put("CLIENT_CODE", new FieldDef("STRING"));
	  this.fields.put("NAME", new FieldDef("STRING"));
	  this.fields.put("IS_FIRST_PERIOD", new FieldDef("BOOLEAN"));
	  String[] _pkFields = {"ID"};
	  this.pkFields = _pkFields;
	  String[] _summaryFields = {};
	  this.summaryFields = _summaryFields;
	  this.queryResultSize = 20;
	}

}
